File: q16-binary-tree-level-order-traverse/q16-sol.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

// BFS = function(root) {
//     res = []; q = [root];
//     while(q.length) {
//         node = q.dequeue();
//         res.push(node.val);
//         if(node.left) q.enqueue;
//         if(node.right) q.enqueue;
//     }
//     return res;
// }

const levelOrder = function(root) {
    
    if (!root) return [];
    
    const ans = [];
    const queue = [root];
    
    while(queue.length) {
        let temp = [];
        let currLen = queue.length;
        
        for (let j = 0; j < currLen; j++) {
            
            let currNode = queue.shift();
            temp.push(currNode.val);
            
            if(currNode.left) 
                queue.push(currNode.left);
               
            if(currNode.right)
                queue.push(currNode.right);
            
        }
        ans.push(temp);
    }
    return ans;
};

// time: O(n)
// space: O(n)

File: ztm-ds-algo/6-binary-trees/q16-binary-tree-level-order-traverse/q16-sol.py
# // LeetCode 102.
# // https://leetcode.com/problems/binary-tree-level-order-traversal/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# BFS method
from collections import deque 

class Solution:
    def levelOrder(self, root):
        if not root:
            return []
        
        ans = []
        q = deque([root])
        
        while len(q) > 0:
            temp = []
            currLen = len(q)
            
            for _ in range(currLen):
                currNode = q.popleft()
                temp.append(currNode.val)
                if currNode.left is not None:
                    q.append(currNode.left)
                if currNode.right is not None:
                    q.append(currNode.right)
                    
            ans.append(temp)
        return ans

# // time: O(n)
# // space: O(n)